sqsdomain/cw_pool.go

```diff
@@ -0,0 +1,65 @@
+package sqsdomain
+
+import "github.com/osmosis-labs/osmosis/osmomath"
+
+// CosmWasm contract info from [cw2 spec](https://github.com/CosmWasm/cw-minus/blob/main/packages/cw2/README.md)
+type ContractInfo struct {
+	Contract string `json:"contract"`
+	Version  string `json:"version"`
+}
+
+// Check if the contract info matches the given contract and version
+// The version can be a semver range
+func (ci *ContractInfo) Matches(contract, version string) bool {
+	return ci.Contract == contract && ci.Version == version
+}
+
+// CosmWasmPoolModel is a model for the pool data of a CosmWasm pool
+// It includes the contract info and the pool data
+// The CWPoolData works like a tagged union to hold different types of data
+// depending on the contract and its version
+type CosmWasmPoolModel struct {
+	ContractInfo ContractInfo `json:"contract_info"`
+	Data         CWPoolData   `json:"data"`
+}
+
+// CWPoolData is the custom data for each type of CosmWasm pool
+// This struct is intended to work like tagged union in other languages
+// so that it can hold different types of data depending on the contract
+type CWPoolData struct {
+	// Data for AlloyTransmuter contract, must be present if and only if `IsAlloyTransmuter()` is true
+	AlloyTransmuter *AlloyTransmuterData `json:"alloy_transmuter,omitempty"`
+}
+
+func NewCWPoolModel(contract string, version string, data CWPoolData) *CosmWasmPoolModel {
+	return &CosmWasmPoolModel{
+		ContractInfo: ContractInfo{
+			Contract: contract,
+			Version:  version,
+		},
+		Data: data,
+	}
+}
+
+func (model *CosmWasmPoolModel) IsAlloyTransmuter() bool {
+	return model.ContractInfo.Matches("crates.io:transmuter", "3.0.0")
```

Comment:
What if the contract is migrated? We would then require code updates and redeployment?

Please consider removing this hard-coded dependency here and on the node side.

The current code ID model allows us to simply update the config without performing any code changes

Reply:
Please consider going back to the code ID config when determining if a pool is an alloy transmuter within SQS. 

To achieve this on the Osmosis node side, we can continue utilizing the CW info but only match the name e.g. `crates.io:transmuter-alloy` (no version). This way, if the version is updated, it will still be auto-propagated from Osmosis node. However, give SQS the flexibility of handling it via standard means.

The proposed code ID approach would ensure compatibility with how other cw pool types are handled while also allowing the flexibility of only updating configs without code changes if migration is done.

With the current approach, we still need a boolean config but also hardcode the version here which I think is suboptimal for operations.

Please let me know what you think

Reply:
For visibility, this is Boss's reply (did not get submitted publicly) that I will address in a separate message

![image](https://github.com/osmosis-labs/sqs/assets/34196718/88772311-ee11-4fc0-812e-6c07eefdf714)

Reply:
I think that we should discuss this decision via a separate ADR to avoid blocking this line of work.

I would like to avoid merging partial progress with a new paradigm that has trade-offs. If we want to change how CW pools are handled, let's aim to change this for all pools at once rather than introducing custom handling for the alloyed pool but keeping everything else via code IDs.

What I would like to avoid is us never coming back to this and then having to deal with operational challenges stemming from the custom logic.

In terms of versions, I don't understand why we need to track these hard-coded versions in general when contract names should suffice. We should establish a rule that all future versions of a cw pool model struct transferred between node and sqs are backward compatible. This removes the need for version matching on the node side, removing the need to either have a config (node-side) or hard code it in `sqsdomain`.

I'm supportive of using versions in the SQS config rather than hardcoding. But again, I propose making that decision in a separate ADR and/or issue to avoid overly long debates on the otherwise complete work here.

Again, my main counter-argument to the proposed change in this PR is the following:
- Operational challenges stemming from hardcoding versions on both osmosis and sqs side
- Applying an operational change to one cw pool only, causing inconsistencies in handling
- The trade-offs are not fully analyzed before making a change that impacts service operations

However, I think that we should explore the version idea in SQS configs in general. That is an interesting proposal.

@iboss-ptk please let me know what you think

Reply:
Fair point on consistency.

I will revert back the config to code ids and and have a separated ADR for versioning we discussed to unblock things. That being said. Since old transmuter is not forward compatible, it still need minimum version `>=3.0.0` hardcoded. Even though it is a non-issue on the sqs side, ingester needs this information to do pool transformation properly as current transmuter v3 on mainnet is already `crates.io:transmuter` `v3.0.0` so this can't be changed to `crates.io:transmuter-alloy` unless having a migration (and I don't think that make much sense anyway since it is still a newer version of the same thing).

Reply:
Thanks

Per [this comment](https://github.com/osmosis-labs/sqs/pull/245#discussion_r1625856060), the latest solution is more extensible.

If we keep it as a convention that all alloyed transmuters stay within the same major version and add a relevant semver constraint (`<4.0.0`), we solve all future extensibility problems.

In the original approach, due to semver being hardcoded directly, we would have to make code changes and dependency updates to all of `sqsdomain`, `osmosis` and `sqs` if a migration was done.

Now, migrations within the same major are auto-accounted for